pilot/pkg/status/reporter.go

```diff
@@ -0,0 +1,309 @@
+			}
+			if out.InProgressResources[key] >= out.DataPlaneCount {
+				// if this resource is done reconciling, let's not worry about it anymore
+				finishedResources = append(finishedResources, res)
+				// deleting it here doesn't work because we have a read lock and are inside an iterator.
+				// TODO: this will leak when a resource never reaches 100% before it is replaced.
+				// TODO: do deletes propagate through this thing?
+			}
+		}
+	}
+	return out, finishedResources
+}
+
+// For efficiency, we don't want to be checking on resources that have already reached 100% distribution.
+// When this happens, we remove them from our watch list.
+func (r *Reporter) removeCompletedResource(completedResources []Resource) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	var toDelete []Resource
+	for _, item := range completedResources {
+		// TODO: handle cache miss
+		total := r.inProgressResources[item.ToModelKey()].completedIterations + 1
+		if int64(total) > (time.Minute.Milliseconds() / r.UpdateInterval.Milliseconds()) {
+			//remove from inProgressResources // TODO: cleanup completedResources
+			toDelete = append(toDelete, item)
+		} else {
+			r.inProgressResources[item.ToModelKey()].completedIterations = total
+		}
+	}
+	for _, resource := range toDelete {
+		delete(r.inProgressResources, resource.ToModelKey())
+	}
+}
+
+// This function must be called every time a resource change is detected by pilot.  This allows us to lookup
+// only the resources we expect to be in flight, not the ones that have already distributed
+func (r *Reporter) AddInProgressResource(res model.Config) {
+	myRes := ResourceFromModelConfig(res)
+	if myRes == nil {
+		scope.Errorf("Unable to locate schema for %v, will not update status.", res)
+		return
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.inProgressResources[myRes.ToModelKey()] = &inProgressEntry{
+		Resource:            *myRes,
+		completedIterations: 0,
+	}
+}
+
+func (r *Reporter) DeleteInProgressResource(res model.Config) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	delete(r.inProgressResources, res.Key())
+}
+
+// generate a distribution report and write it to a ConfigMap for the leader to read.
+func (r *Reporter) writeReport(ctx context.Context) {
+	report, finishedResources := r.buildReport()
+	go r.removeCompletedResource(finishedResources)
+	//write to kubernetes here.
+	reportbytes, err := yaml.Marshal(report)
+	if err != nil {
+		scope.Errorf("Error serializing Distribution Report: %v", err)
+		return
+	}
+	r.cm.Data[dataField] = string(reportbytes)
+	// TODO: short circuit this write in the leader
+	_, err = CreateOrUpdateConfigMap(ctx, r.cm, r.client)
+	if err != nil {
+		scope.Errorf("Error writing Distribution Report: %v", err)
+	}
+}
+
+// this is lifted with few modifications from kubeadm's apiclient
+func CreateOrUpdateConfigMap(ctx context.Context, cm *corev1.ConfigMap, client v1.ConfigMapInterface) (res *corev1.ConfigMap, err error) {
```

Comment:
the configmap is solely owned by one istiod, and writes are always serialized right? So we don't need to worry about two writes occurring in parallel?

Reply:
if not - we should handle conflicts probably

Reply:
This is a blocking call in a synchronous for loop, so I think we should be safe.